web/Services/LandlordService.cs

```diff
@@ -100,4 +113,49 @@ public async Task ApproveLandlord(int landlordId,  BricksAndHeartsUser user)
         landlord.ApprovalAdminId = user.Id;
         await _dbContext.SaveChangesAsync();
     }
+
+    public LandlordDbModel? FindLandlordWithInviteLink(string inviteLink)
+    {
+        return _dbContext.Landlords.SingleOrDefault(l => l.InviteLink == inviteLink);
+    }
+    
+    // Link existing landlord with user
+    public async Task<ILandlordService.LinkUserWithLandlordResult> LinkExistingLandlordWithUser(
+        string inviteLink,
+        BricksAndHeartsUser user)
+    {
+        // We want to atomically update multiple records (insert a landlord, then set the user's landlord id), so first start a transaction
+        var transaction = await _dbContext.Database.BeginTransactionAsync();
+        await using (transaction)
+        {
+            // Check that the link exists
+            var landlord = _dbContext.Landlords.SingleOrDefault(l => l.InviteLink == inviteLink);
+            if (landlord == null)
+            {
+                return ILandlordService.LinkUserWithLandlordResult.ErrorLinkDoesNotExist;
+            }
+
+            // Check the user doesn't already have a landlord associated
+            var userRecord = _dbContext.Users.Single(u => u.Id == user.Id);
+            if (userRecord.LandlordId != null)
+            {
+                return ILandlordService.LinkUserWithLandlordResult.ErrorUserAlreadyHasLandlordRecord;
+            }
+
+            // Disable invite link
+            landlord.InviteLink = null;
+            landlord.User = userRecord;
+            await _dbContext.SaveChangesAsync();
+
+            // Now we can update the user record too
+            userRecord.LandlordId = landlord.Id;
+            userRecord.Landlord = landlord;
+            user.LandlordId = landlord.Id;
+            await _dbContext.SaveChangesAsync();
+
+            // and finally commit
+            await transaction.CommitAsync();

```

Comment:
This is mostly my fault for misleading comments/code elsewhere. We don't need two calls to SaveChanges here. We do still need the transaction to make sure someone else isn't using the invite link at the same time as us - and we need `REPEATABLE READ` for that, since we need to lock the landlord record.

In the database the only way to link landlords to users is to set user.LandlordId. Entity Framework should translate all of these into the same thing, so we only need one

* `landlord.User = userRecord;`
* `userRecord.LandlordId = landlord.Id;`
* `user.LandlordId = landlord.Id;`

```suggestion
            landlord.InviteLink = null;
            userRecord.LandlordId = landlord.Id;
            await _dbContext.SaveChangesAsync();
            await transaction.CommitAsync();

            user.LandlordId = landlord.Id; // Update the in memory user object that `GetCurrentUser()` returns
```

Reply:
ok :-)